File: unilingo_frontend/src/api/leaderboard.ts
/**
 * Leaderboard API service
 */
import apiClient from './client';
import { getCached, makeCacheKey } from './cache';

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  username: string | null;
  full_name: string;
  avatar_url: string | null;
  avg_band_score: number;
  total_tests: number;
  total_xp: number;
}

export interface LeaderboardData {
  period: string;
  entries: LeaderboardEntry[];
  my_rank: LeaderboardEntry;
}

export const leaderboardAPI = {
  get: async (period: 'weekly' | 'monthly' | 'all_time' = 'weekly'): Promise<LeaderboardData> => {
    const params = { period };
    return getCached(makeCacheKey('leaderboard:list', params), async () => {
      const { data } = await apiClient.get('/leaderboard', { params });
      return data;
    }, 30_000);
  },

  getMyRank: async (period: string = 'weekly') => {
    const params = { period };
    return getCached(makeCacheKey('leaderboard:me', params), async () => {
      const { data } = await apiClient.get('/leaderboard/me', { params });
      return data;
    }, 30_000);
  },
};

File: unilingo-backend/app/schemas/leaderboard.py
"""
Pydantic schemas for Leaderboard & Notifications
"""
import uuid
from datetime import time
from pydantic import BaseModel


# ─── Leaderboard ───

class LeaderboardEntry(BaseModel):
    rank: int
    user_id: uuid.UUID
    username: str | None
    full_name: str
    avatar_url: str | None
    avg_band_score: float
    total_tests: int
    total_xp: int

    model_config = {"from_attributes": True}


class LeaderboardResponse(BaseModel):
    period: str  # "weekly", "monthly", "all_time"
    entries: list[LeaderboardEntry]
    my_rank: LeaderboardEntry | None


# ─── Notifications ───

class NotificationSettingsResponse(BaseModel):
    daily_reminder: bool
    reminder_time: time
    new_words_reminder: bool
    streak_reminder: bool
    leaderboard_update: bool

    model_config = {"from_attributes": True}


class UpdateNotificationSettingsRequest(BaseModel):
    daily_reminder: bool | None = None
    reminder_time: time | None = None
    new_words_reminder: bool | None = None
    streak_reminder: bool | None = None
    leaderboard_update: bool | None = None


class RegisterDeviceRequest(BaseModel):
    fcm_token: str
    device_type: str  # "ios" or "android"
    device_name: str | None = None


class NotificationItem(BaseModel):
    id: uuid.UUID
    title: str
    body: str
    type: str
    data: dict | None
    is_read: bool
    created_at: str
